### ISSUE TITLE
Tooltip

### ISSUE THREAD

**Issue by BobbieBarker** [ng-bootstrap/core#8]
Recreate BS3 Tooltip with ng2 component.

**Comment by RobJacobs** [comment]
We're going to need a positioning service first.  May want to consider a service for managing the triggers rather than have that logic in the tooltip component.  May want to consider a distinct and separate component from the popover component as what is in the existing library is unwieldy.

**Comment by pkozlowski-opensource** [comment]
Here are some notes on use-cases / API / implementation details: https://docs.google.com/document/d/1mRTbpxSNg1dLfK-d_UIns8d1SKuQvcO3B7k-A_T_VIU/edit?usp=sharing

**Comment by pkozlowski-opensource** [comment]
@ng-boostrap/core I'm starting to implement tooltips / popovers and I would need your opinion on one part of the API: triggers. The part I'm struggling with is flexibility vs. ease of use. More precisely let's assume that we want to support custom triggers (both close and open). In such case the simplest / most flexible approach is to expose the tooltip directive so people can do:

```javascript
<div ngbTooltip=”Some text” #myTip=”ngbTooltip” (mouseenter)=”myTip.open()” (mouseleave)=”myTip.close()”>
```

Using the above people can subscribe to any events they wish. But what to do about default triggers (mouse enter + leave)? Do we want to provide some shortcuts / defaults? Sth like:

```javascript
<div ngbTooltip=”Some text” ngbTooltipDefaultTriggers>
```

or 


```javascript
<div ngbTooltip=”Some text” ngbTooltipTriggers="click:mouseleave">
```

I'm not big fan of supporting `<div ngbTooltip=”Some text” ngbTooltipTriggers="click:mouseleave">` form (at least not initially) since it isn't much shorter as compared to `<div ngbTooltip=”Some text” #myTip=”ngbTooltip” (mouseenter)=”myTip.open()” (click)=”myTip.close()”>` and a corresponding directive could be done in the user-land.

What is your opinion? What kind of API you would like to see / use?

**Comment by Foxandxss** [comment]
I would like to have the default behaviour without the need of writting any attribute, just

```
<div ngbTooltip="Foo">
```

**Comment by pkozlowski-opensource** [comment]
@Foxandxss sure, we would all like to have it. But I don't know how to do it technically speaking. That is, have default triggers and flexibility for custom ones. The problem is that if you add default triggers in the tooltip directive then supporting custom ones get harder.

Alternatively we would have to do something like this for custom triggers:

`<div ngbTooltip=”Some text” #myTip=”ngbTooltip” (mouseenter)=”myTip.open()” (click)=”myTip.close()” ngbTooltipNoDefaultTriggers>` 

But even in this case we couldn't avoid registration of even listeners.

Based on the above, do you have idea how to implement API that would cover your proposal and custom triggers?

**Comment by Foxandxss** [comment]
Can't we have something like:

```
<div ngbTooltip="Foo"></div>
```

That will use the default triggers, click and mouseleave.

Then if we decide we want custom triggers, we can use:

```
<div ngbTooltip="Foo" triggers="mouseenter:mouseleave"></div>
```

That form or the separate form you mentioned earlier.

The big point is, why can't we override a default behaviour without having to add any attribute that say "we don't want default behavior"

**Comment by pkozlowski-opensource** [comment]
> Then if we decide we want custom triggers, we can use `<div ngbTooltip="Foo" triggers="mouseenter:mouseleave"></div>`

We can do it, technically speaking, but this is _precisely_ the API I would like to avoid as far more flexible option that doesn't require any implementation is: `<div ngbTooltip=”Some text” #myTip=”ngbTooltip” (mouseenter)=”myTip.open()” (click)=”myTip.close()”>`

But if we use this approach then we don't have any attribute that we could test to know if we should add default triggers or not. This is the crux!

In short we can either has nice syntax for default triggers or nice solution for custom triggers. At the moment I'm failing to see how we could have nice syntax for both scenarios. And this is where I need your help :-)

**Comment by RobJacobs** [comment]
I like the idea of subscribing to the events explicitly.  With the current api, folks struggle when straying from the default triggers offered and defining multiple triggers for open or close events.  If we do go with the second approach, I would like to have the ability to set the open and close triggers like so:

    <div ngbTooltip="foo" triggers="{open: 'focus mouseenter', close: 'blur mouseleave'}"><div>

Or maybe:

    <div ngbTooltip="foo" openTriggers="focus mouseenter" closeTriggers="blur mouseleave"><div>

**Comment by pkozlowski-opensource** [comment]
As I've said I would very much prefer steering clear from the triggers API we've got today since with ng2 we can do much better and get more flexibility without doing anything: `<div ngbTooltip=”Some text” #myTip=”ngbTooltip” (mouseenter)=”myTip.open()” (click)=”myTip.close()”>`

**Comment by RobJacobs** [comment]
And I'm assuming folks could create their own shortcut directives if they would rather not set the events inline?

**Comment by pkozlowski-opensource** [comment]
> And I'm assuming folks could create their own shortcut directives if they would rather not set the events inline?

Yes, it would be trivial in the user-land.

The only thing that worries me that users might be put off by the amount of typing / verbosity of a typical use case....

**Comment by icfantv** [comment]
Have we looked at what material is doing?  Have they implemented this widget yet?

What if we just don't have any triggers at all and explicitly force the user to add their own to the directive element?  Or is there a way to say "this is my open trigger(s)" and "this is my close trigger(s)?"

**Comment by pkozlowski-opensource** [comment]
@icfantv didn't look at what material did

> What if we just don't have any triggers at all and explicitly force the user to add their own to the directive element? Or is there a way to say "this is my open trigger(s)" and "this is my close trigger(s)?"

Could you be more precise? I'm trying to focus the discussion on exact API proposals for _both_ default and custom triggers.

**Comment by icfantv** [comment]
I haven't dived into components yet outside of what we covered at ng-conf so I don't know if what I'm saying is even valid.  It almost sounds like you're suggesting multiple templates based on whether or not the user wants anything but the default triggers.

Because templates are pre compiled, I'm assuming there's not a way to just define an open trigger attribute and a close trigger attribute and let the user define their own open and close events letting us bind to them later?  We could also simply document that they need to bind their own events to the element and we can just document all the available ones...

Do we still want to support manual triggers?  Like `is-open`?

**Comment by mendeza** [comment]
Just to come at this discussion from an entirely different direction, isn't Bootstrap still mobile-first? Should `mouseenter` and `mouseleave` really be the defaults? Any thought given to [Pointer Events](https://www.w3.org/TR/pointerevents/)? To be frank, the mobile experience of the AngularUI 1.x tooltip and popover has been rather poor. Would't we like to do better in this project?

**Comment by pkozlowski-opensource** [comment]
@mendeza you've got a point about the fact that "defaults" will depend on a platform. Which would be one more reason for not hard-coding defaults in the directive.

Having said this I think that tooltips / popovers are bad for mobile anyway.

**Comment by mendeza** [comment]
@pkozlowski-opensource - I agree. But many consumers will have a single UI for both desktop and mobile; after all that is one of the core missions of Bootstrap as a 
>...responsive, mobile-first projects on the web.

**Comment by pkozlowski-opensource** [comment]
> But many consumers will have a single UI for both desktop and mobile;

Yeh, this is the sad reality I see as well :-/ Anyway, this is an important point in the "default triggers" discussion. @mendeza what would be your API proposal here?

**Comment by mendeza** [comment]
Browser support for JS pointer-events is [terrible](http://caniuse.com/#search=pointerenter). I suspect you have no interest in freighting ng-bootstrap with polyfills for this - can be left up to consumer choice. My quick from-the-hip impression is that I like @RobJacobs suggestion.

**Comment by pkozlowski-opensource** [comment]
OK, so after much deliberation my plan is to use a dedicated `triggers` input (attribute). This is how it would work:

* `<div ngbTooltip="Killer tip!">` - for a default / standard case
* `<div ngbTooltip="Killer tip!" triggers="click:mouseout">` - for non-default triggers
* `<div ngbTooltip="Killer tip!" triggers="click:mouseout focus">` - for non-default, multiple triggers
* `<div ngbTooltip="Killer tip!" triggers="manual" #t="ngbTooltip" (click)="t.open()" (document:keydown.esc)="t.open()">` - for fine grained / I want to go crazy control

The above proposal surrenders to the fact that user ergonomic is more important that ease of implementation for us :-)

**Comment by Foxandxss** [comment]
I love it.

**Comment by mendeza** [comment]
> The above proposal surrenders to the fact that user ergonomic is more important that ease of implementation for us :-)

@pkozlowski-opensource - there's a term for this, right? I believe it's known as "API UX." I'll run this by my team.

**Comment by icfantv** [comment]
@pkozlowski-opensource this is great, i really like it as well.

**Comment by pkozlowski-opensource** [comment]
Blocked on https://github.com/angular/angular/issues/9293 and https://github.com/angular/angular/issues/8617
